ledger/accountdb.go

```diff
@@ -2682,10 +2688,37 @@ func (qs *accountsDbQueries) lookupKeyValue(key string) (pv persistedKVData, err
 	return
 }
 
+// keyPrefixIntervalPreprocessing is implemented to generate an interval for DB query: lookupKeysByPrefixStmt
+// Such DB query was designed this way, to trigger the binary search optimization in SQLITE3.
+// The DB comparison for blob typed primary key is lexicographic, i.e., byte by byte.
+// In this way, we can introduce an interval that a primary key should be >= some prefix, < some prefix increment.
+// A corner case to consider is that, the prefix has last byte 0xFF, or the prefix is full of 0xFF.
+// The first case can be solved by carrying, e.g., prefix = 0x1EFF -> 0x1F00
+// The second case can be solved by disregarding the upper limit, i.e., prefix = 0xFFFF -> interval being >= 0xFF
+func keyPrefixIntervalPreprocessing(prefix []byte) ([]byte, []byte) {
```

Comment:
No, I believe that's the right value.

Reply:
actually good find, for `prefixIncr` we do expect increment by 1 like normal integers, but only exception is last full 0xFF.

Thus, the interval should have form [{0xFE, 0xFF}, {0xFF, 0x00}], or [{0xFF, 0xFF}, nil], or [{}, nil], if prefix is empty. The only question here is, when `nil` is passed into this function, I think the right result should be [{}, nil], tho only call is from `lookupKeysByPrefix`, which guarantees no nil being input.

Reply:
Decide to handle the `nil` case too, pushing.

Reply:
I like Zeph's suggestions to test just this function so that we can have drop dead simple cases, but I still think the code is right as is.  Can we discuss the case Zeph brought up?  The case is:
```
1.4. prefix --> prefixBytes, prefixIncrBytes := [255 255 52 86] --> [255 255 52 86], [255 255 52 87]
```
I believe that is correct.  If you want all the byte slices that start with 255,255,52,86 then you want anything that comes after 255,255,52,86 but before 255,255,52,87.

Zeph's proposed increment, 255,255,53,86 as an upper bound would, for example, return 255,255,53,50. But that does not have a 255,255,52,86 prefix.

Reply:
I believe such case, namely 
```
1.4. prefix --> prefixBytes, prefixIncrBytes := [255 255 52 86] --> [255 255 52 86], [255 255 52 87]
```
is correct. The conclusion we had was finding everything after `[255 255 52, 86]`, but before `[255 255 53 87]`.

Zeph's increment by one does not apply to least significant bit, but seems applying to most significant bit, if possible.

This gives me some thinking about if I should change naming of this function.

Reply:
So let me try and understand. Suppose we would like to find all byte strings that are prefixed by `255 255 52 86`.
If we could use a regex it would like like `255 255 52 86 N*` where `N` denotes any byte.

The constraint `x >= 255 255 52 86` gets us part of the way as certainly `255 255 52 86 N* >= 255 255 52 86`.

However, lexicographically (reading left to right), we also have:

```
255 255 53 > 255 255 52 86
```
(e.g. `zze` comes after `zzdg` in the dictionary of non-existent words starting with z)

But we certainly do not want to include `255 255 53` in our result set, as it isn't prefixed by `255 255 52 86`. To achieve this exclusion, we should add another inequality and obtain our final constraint set:

```
x >= 255 255 52 86 AND x < 255 255 53
```

We still include strings like `255 255 52 86 255 255 255` since it is `< 255 255 53`

Reply:
In your example, there's no reason to carry. Just increment the last byte (86) to 87.

Reply:
edited the above...

Reply:
I should at least address the other implied constraint present in this PR:
```
x >= 255 255 52 86 AND x < 255 255 52 87
```
it seems to work as well in this case. So I need to think a little more. Are these equivalent? Is there a counterexample? Is there a special case where one increment from left to right works but right to left doesn't (or the opposite) ?

Reply:
But now that I modded the constraint to `< 255 255 53 86`, unfortunately `255 255 53` has slipped back in. It is still exluded by `< 255 255 52 87`, and so it does seem like the original approach is the correct one.

Reply:
Think about it with a normal paper dictionary and english words.  If you wanted all the words that start with "hot", you'd want to skip ahead to "hot" with a binary search (that's what we're trying to convince sqlite to do) and then scan forward.  When do you stop? As soon as you hit a word that doesn't start with "hot". If we had a fake new letter of the alphabet, `z'`, we could go until we are past "hotz`", but since we don't, we pick the first letter that comes after t, and go until we pass the word "hou".

That work fine, unless we want the words that start with "buzz". We can't rollover the "z", so we have to percolate up, and stop at "bvaa".

Reply:
resolving as I think we've reduced this down to the case of whether we need to erase 0's after incrementing the key for obtaining the upper bound.